contracts/optimistic-ethereum/OVM/verification/OVM_StateTransitioner.sol

```diff
@@ -173,81 +182,62 @@ contract OVM_StateTransitioner is Lib_AddressResolver, OVM_FraudContributor, iOV
         require(
-            ovmStateManager.hasAccount(_ovmContractAddress) == false,
-            "Account state has already been proven"
-        );
-
-        require(
-            _account.codeHash == Lib_EthUtils.getCodeHash(_ethContractAddress),
-            "Invalid code hash provided."
-        );
-
-        require(
-            Lib_SecureMerkleTrie.verifyInclusionProof(
-                abi.encodePacked(_ovmContractAddress),
-                Lib_OVMCodec.encodeEVMAccount(_account),
-                _stateTrieWitness,
-                preStateRoot
+            (
+                ovmStateManager.hasAccount(_ovmContractAddress) == false
+                && ovmStateManager.hasEmptyAccount(_ovmContractAddress) == false
             ),
-            "Account state is not correct or invalid inclusion proof provided."
-        );
-
-        ovmStateManager.putAccount(
-            _ovmContractAddress,
-            Lib_OVMCodec.Account({
-                nonce: _account.nonce,
-                balance: _account.balance,
-                storageRoot: _account.storageRoot,
-                codeHash: _account.codeHash,
-                ethAddress: _ethContractAddress,
-                isFresh: false
-            })
-        );
-    }
-
-    /**
-     * Allows a user to prove that an account does *not* exist in the state.
-     * @param _ovmContractAddress Address of the contract on the OVM.
-     * @param _stateTrieWitness Proof of the (empty) account state.
-     */
-    function proveEmptyContractState(
-        address _ovmContractAddress,
-        bytes memory _stateTrieWitness
-    )
-        override
-        public
-        onlyDuringPhase(TransitionPhase.PRE_EXECUTION)
-        contributesToFraudProof(preStateRoot, transactionHash)
-    {
-        // Exit quickly to avoid unnecessary work.
-        require(
-            ovmStateManager.hasEmptyAccount(_ovmContractAddress) == false,
             "Account state has already been proven."
         );
 
-        require(
-            Lib_SecureMerkleTrie.verifyExclusionProof(
-                abi.encodePacked(_ovmContractAddress),
-                _stateTrieWitness,
-                preStateRoot
-            ),
-            "Account is not empty or invalid inclusion proof provided."
+        (
+            bool exists,
+            bytes memory encodedAccount
+        ) = Lib_SecureMerkleTrie.get(
+            abi.encodePacked(_ovmContractAddress),
+            _stateTrieWitness,
+            preStateRoot
         );
 
-        ovmStateManager.putEmptyAccount(_ovmContractAddress);
+        if (exists == true) {
+            Lib_OVMCodec.EVMAccount memory account = Lib_OVMCodec.decodeEVMAccount(
+                encodedAccount
+            );
+
+            address ethContractAddress = _ethContractAddress;
+            if (account.codeHash == EMPTY_ACCOUNT_CODE_HASH) {
+                ethContractAddress = 0x0000c0De0000C0DE0000c0de0000C0DE0000c0De;
```

Comment:
Use a known empty account to avoid an attack in which the user deploys code to the contract after supplying it here.

Reply:
Should some of these comments be in the code itself?

Reply:
Yeah, sorry. Am cleaning up this PR as I fix the tests.

Reply:
Recommend to set this to ZERO_ADDRESS--less fun, but I fear this may be confusing, especially since we have a similar use of `c0dec` elsewhere IIRC.

Reply:
Fixed.